cumulus/parachains/runtimes/coretime/coretime-rococo/src/coretime.rs

```diff
@@ -0,0 +1,298 @@
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: Weight::from_parts(1000000000, 200000),
+				call: request_revenue_info_at_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, MultiLocation::parent(), message.clone()) {
+			Ok(_) => log::info!(
+				target: "runtime::coretime",
+				"Request for revenue information sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Request for revenue information failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn credit_account(who: Self::AccountId, amount: Self::Balance) {
+		use crate::coretime::CoretimeProviderCalls::CreditAccount;
+		let credit_account_call = BrokerRuntimePallets::<
+			CoreIndex,
+			BlockNumber,
+			AccountId,
+			Balance,
+			CoreAssignment,
+			PartsOf57600,
+		>::Broker(CreditAccount(who, amount));
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: Weight::from_parts(1000000000, 200000),
+				call: credit_account_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, MultiLocation::parent(), message.clone()) {
+			Ok(_) => log::info!(
+				target: "runtime::coretime",
+				"Instruction to credit account sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Instruction to credit account failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn assign_core(
+		core: CoreIndex,
+		begin: Self::BlockNumber,
+		assignment: Vec<(CoreAssignment, PartsOf57600)>,
+		end_hint: Option<Self::BlockNumber>,
+	) {
+		use crate::coretime::CoretimeProviderCalls::AssignCore;
+		let assign_core_call = BrokerRuntimePallets::<
+			CoreIndex,
+			BlockNumber,
+			AccountId,
+			Balance,
+			CoreAssignment,
+			PartsOf57600,
+		>::Broker(AssignCore(core, begin, assignment, end_hint));
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: Weight::from_parts(1000000000, 200000),
+				call: assign_core_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, MultiLocation::parent(), message.clone()) {
+			Ok(_) => log::info!(
+				target: "runtime::coretime",
+				"Core assignment sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Core assignment failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn check_notify_core_count() -> Option<u16> {
+		let count = CoreCount::get();
+		CoreCount::set(&None);
+		count
+	}
+
+	fn check_notify_revenue_info() -> Option<(Self::BlockNumber, Self::Balance)> {
+		let revenue = CoretimeRevenue::get();
+		CoretimeRevenue::set(&None);
+		revenue
+	}
```

Comment:
This functions should return what `request_core_info` and `request_revenue_info_at` "return" from the relay chain.

Generally I think that this isn't correctly implemented. I'm also not sure why the relay chain is not just calling some function inside the broker pallet to inform about the result of the request.

Reply:
The extra extrinsic that Rob and I discussed will set `CoreCount` when it's called from the XCM from the relay chain, and another similar extrinsic will be added for `CoretimeRevenue`. So if these two are then called they will return the value that was returned from the relay.

This may be part of the interface that just doesn't make sense to implement on the parachain side in the way the docs describe, this made the most sense to me but I wasn't sure if it would ever be called.

What do you think:
1. Keep these implementations and create the new extrinsics.
2. Remove these two implementations for the parachain (docs explicitly say they're for the relay) and just call a function inside broker as you've suggested.

Reply:
> This may be part of the interface that just doesn't make sense to implement on the parachain side in the way the docs describe, this made the most sense to me but I wasn't sure if it would ever be called.

I don't get what you are saying here. Reading the docs these functions are for getting the results of calls that we send to the relay chain.

Reply:
Yea sorry I think I see where I was going wrong now - I thought that it was referring to this interface on the relay chain side, and thought that these two were actually sort of redundant on the parachain due to the message that was sent back after assign core.
But actually we had just misinterpreted the comment for `assign_core` - Rob and I thought that when the relay chain assigned the core they were meant to respond with an XCM Transact that would need to call an extrinsic that doesn't exist to just update the parachain storage, but actually what you've just pointed out is that they don't send anything, they change their state and wait for us to ask them, in which case we use this interface.

Is that what you mean? I think I know what way to implement these properly now.

Reply:
> Rob and I thought that when the relay chain assigned the core they were meant to respond with an XCM Transact that would need to call an extrinsic that doesn't exist to just update the parachain storage, but actually what you've just pointed out is that they don't send anything, they change their state and wait for us to ask them, in which case we use this interface.

You need to communicate the result. However you do it. You can do this by sending a XCM as you said or by doing a storage proof.